src/wasm-interpreter.h

```diff
@@ -1174,28 +1246,167 @@ class ExpressionRunner : public OverriddenVisitor<SubType, Flow> {
     assert(flow.values.size() > curr->index);
     return Flow(flow.values[curr->index]);
   }
+  Flow visitLocalGet(LocalGet* curr) {
+    NOTE_ENTER("LocalGet");
+    NOTE_EVAL1(curr->index);
+    // Check if a constant value has been set in the context of this runner.
+    auto iter = localValues.find(curr->index);
+    if (iter != localValues.end()) {
+      return Flow(std::move(iter->second));
+    }
+    return Flow(NONCONSTANT_FLOW);
+  }
+  Flow visitLocalSet(LocalSet* curr) {
+    NOTE_ENTER("LocalSet");
+    NOTE_EVAL1(curr->index);
+    if (!(flags & FlagValues::PRESERVE_SIDEEFFECTS)) {
+      // If we are evaluating and not replacing the expression, see if there is
+      // a value flowing through a tee.
+      if (curr->type.isConcrete()) {
+        assert(curr->isTee());
+        return visit(curr->value);
+      }
+      // Otherwise remember the constant value set, if any, for subsequent gets.
+      if (setLocalValue(curr->index, curr->value)) {
```

Comment:
The code around here would be simpler, I think, if we computed `visit(curr->value)` unconditionally first, and then decided what to return later. The code right not hides the computation inside the call to `setLocalValue`, which is less obvious.

Do we even need `setLocalValue(index, expression)`? Can't it always take an index and a value?

Reply:
The one place where an expression is useful is in the JS API, where dealing with literals is somewhat an internal mechanism not directly exposed via the API. Moved this to the C API only now, so the parent runner doesn't need the expression variant :)